src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Specialized;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http;
+using System.Runtime.Caching;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.ServiceHub.Framework.Services;
+using NuGet.Common;
+using NuGet.PackageManagement.Telemetry;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.VisualStudio.Internal.Contracts;
+using NuGet.VisualStudio.Telemetry;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public sealed class NuGetPackageFileService : INuGetPackageFileService
+    {
+        public static readonly string IconPrefix = "icon:";
+        public static readonly string LicensePrefix = "license:";
+
+        private ServiceActivationOptions? _options;
+        private IServiceBroker _serviceBroker;
+        private AuthorizationServiceClient? _authorizationServiceClient;
+        private INuGetTelemetryProvider _nuGetTelemetryProvider;
+        private bool _disposedValue;
+
+        internal readonly static MemoryCache IdentityToUriCache = new MemoryCache("PackageSearchMetadata",
+            new NameValueCollection
+            {
+                { "cacheMemoryLimitMegabytes", "4" },
+                { "physicalMemoryLimitPercentage", "0" },
+                { "pollingInterval", "00:02:00" }
+            });
+
+        private static readonly CacheItemPolicy CacheItemPolicy = new CacheItemPolicy
+        {
+            SlidingExpiration = ObjectCache.NoSlidingExpiration,
+            AbsoluteExpiration = ObjectCache.InfiniteAbsoluteExpiration,
+        };
+
+        public static void AddIconToCache(PackageIdentity packageIdentity, Uri iconUri)
+        {
+            string key = NuGetPackageFileService.IconPrefix + packageIdentity.ToString();
```

Comment:
Is this enough? What about source?

Reply:
discussed in the spec here: https://github.com/NuGet/Home/pull/10460/files#diff-245d4e560471354abefcccfae6f6cd50f536202f2476410108e3cdfe53c253cdR108

which currently says: We also chose to only key off of PackageIdentity, instead of PackageIdentity and a package source...staying consistent with our treatment of unique package identities as the same, regardless of where they come from.